### DISCUSSION TITLE
Partitioned writes should be configurable to account for data distribution

### DISCUSSION THREAD

**Discussion by avyfain** [duckdb/duckdb#9555] [Ideas]
When users write out datasets where data is not uniformly distributed within its partitions, they end up with tiny files on some partitions and the same number of very big files in others. Ideally when there isn’t that much data in a partition, duckdb should be able to write out less files.

The documentation states that:
> Currently, one file is written per thread to each directory.

and

> Writing data into many small partitions is expensive. It is generally recommended to have at least 100MB of data per partition.

There could be a way to bottleneck this internally by using internal statistics upfront to estimate the output size and the correct number of threads to assign to the task. Alternatively there could be a compaction step within the process to dynamically union small files after the write.

Inspired by a conversation about this in the dbt Slack:
https://getdbt.slack.com/archives/C039D1J1LA2/p1698770961997609

**Comment by l1t1** [comment]
I have a similar question. 
https://github.com/duckdb/duckdb/discussions/9493
